operators/constellation-node-operator/api/v1alpha1/nodeversion_types.go

```diff
@@ -21,6 +21,8 @@ type NodeVersionSpec struct {
 	KubernetesComponentsReference string `json:"kubernetesComponentsReference,omitempty"`
 	// KubernetesClusterVersion is the advertised Kubernetes version of the cluster.
 	KubernetesClusterVersion string `json:"kubernetesClusterVersion,omitempty"`
+	// MaxNodeBudget is the maximum number of nodes that can be created simultaneously.
+	MaxNodeBudget uint32 `json:"maxNodeBudget,omitempty"`
```

Comment:
I'm against making this a user-facing feature for now, because we did not discuss its semantics sufficiently. What if the user sets it to 1000 - do we replace all control-plane nodes at once? Should we have different budgets for control planes and for workers? Could this be relative? Should we rather implement a different upgrade algorithm ([3533](https://dev.azure.com/Edgeless/Edgeless/_workitems/edit/3533) etc)? I'm also reminded of the evolution of https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#rollingupdatedeployment-v1-apps. 

Afaict this PR would be much smaller if we removed this feature and tried to find a different way to test it.

Reply:
> do we replace all control-plane nodes at once?

Yes, I also tested it setting the budget to 5 when I had 3:2 nodes. But "at-once" only applies to the world how the operator sees it. The join service still forbids multiple control-plane nodes joining at the same time. Note that this is the scenario right after initializing a Constellation with >=3 control planes. Also replace means, adding the nodes first. So in theory you could go from 3 control planes to 6 since the operator only removes a node once the hand over is finished.


> Should we have different budgets for control planes and for workers? 

We might do that in the future, if a customer requires it or we think we need it.


> Could this be relative?

I assume as in a percentage value. Sure, but this is more difficult/complex then setting the number of nodes.


> Should we rather implement a different upgrade algorithm

I think the bug is orthogonal to this proposal, since I'm not reworking the operator replacement algorithm, which would be a large undertaking in my opinion.


> I'm also reminded of the evolution of https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#rollingupdatedeployment-v1-apps

I don't know the evolution, is there a summary somewhere of what happened?


Note that also the operator API is on version v1alpha1, so in my opinion, we don't have to provide any API stability guarantees between Constellation versions and we can completely change the whole upgrade process and APIs between Constellation versions. 

Some of the current fields are not really user-facing in a sense that the user should directly use them. Changing the image reference requires 1. the image reference to be one of our images references 2. the measurements in the join config to match the image. Changing the k8s version requires a config map under that name and for the Constellation to upgrade _correctly_ it has to contain the right set of components and patches.


> Afaict this PR would be much smaller if we removed this feature and tried to find a different way to test it.

Then I'll have another try at the test for this, but this might take a bit of time before I get back to this.